pkg/operator/controller/care/add.go

```diff
@@ -0,0 +1,129 @@
+)
+
+// ControllerName is the name of this controller.
+const ControllerName = "garden-care"
+
+// AddToManager adds Reconciler to the given manager.
+func (r *Reconciler) AddToManager(mgr manager.Manager) error {
+	if r.RuntimeClient == nil {
+		r.RuntimeClient = mgr.GetClient()
+	}
+	if r.Clock == nil {
+		r.Clock = clock.RealClock{}
+	}
+	if r.GardenNamespace == "" {
+		r.GardenNamespace = v1beta1constants.GardenNamespace
+	}
+
+	c, err := builder.
+		ControllerManagedBy(mgr).
+		Named(ControllerName).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: 1,
+			// if going into exponential backoff, wait at most the configured sync period
+			RateLimiter: workqueue.NewWithMaxWaitRateLimiter(workqueue.DefaultControllerRateLimiter(), careSyncPeriod),
+		}).
+		Watches(
+			source.NewKindWithCache(&operatorv1alpha1.Garden{}, mgr.GetCache()),
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(
+				r.GardenPredicate()),
+		).Build(r)
+	if err != nil {
+		return err
+	}
+
+	return c.Watch(
+		source.NewKindWithCache(&resourcesv1alpha1.ManagedResource{}, mgr.GetCache()),
+		mapper.EnqueueRequestsFrom(mapper.MapFunc(r.MapManagedResourceToGarden), mapper.UpdateWithNew, c.GetLogger()),
+		predicateutils.ManagedResourceConditionsChanged(),
+	)
+}
+
+// GardenPredicate is a predicate which returns 'true' for create events, and for update events in case the garden was
+// successfully reconciled.
+func (r *Reconciler) GardenPredicate() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(event.CreateEvent) bool {
+			return true
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			garden, ok := e.ObjectNew.(*operatorv1alpha1.Garden)
+			if !ok {
+				return false
+			}
+
+			oldGarden, ok := e.ObjectOld.(*operatorv1alpha1.Garden)
+			if !ok {
+				return false
+			}
+
+			return gardenReconciledSuccessfully(oldGarden, garden)
+		},
+		DeleteFunc:  func(event.DeleteEvent) bool { return false },
+		GenericFunc: func(event.GenericEvent) bool { return false },
+	}
+}
+
+func gardenReconciledSuccessfully(oldSeed, newSeed *operatorv1alpha1.Garden) bool {
+	oldReconcileCondition := v1beta1helper.GetCondition(oldSeed.Status.Conditions, operatorv1alpha1.GardenReconciled)
+	newReconcileCondition := v1beta1helper.GetCondition(newSeed.Status.Conditions, operatorv1alpha1.GardenReconciled)
+
+	return newReconcileCondition != nil &&
+		newReconcileCondition.Status == gardencorev1beta1.ConditionTrue &&
+		(oldReconcileCondition == nil || oldReconcileCondition.Status != gardencorev1beta1.ConditionTrue)
+}
+
+// MapManagedResourceToGarden is a mapper.MapFunc for mapping a ManagedResource to the owning Garden.
+func (r *Reconciler) MapManagedResourceToGarden(ctx context.Context, log logr.Logger, _ client.Reader, _ client.Object) []reconcile.Request {
+	gardenList := &operatorv1alpha1.GardenList{}
+	if err := r.RuntimeClient.List(ctx, gardenList); err != nil {
```

Comment:
metadata-only call

Reply:
We can in addition save the name of the garden after the first list and always return:

```
return []reconcile.Request{{NamespacedName: types.NamespacedName{Name: r.GardenName}}}
```

Reply:
> We can in addition save the name of the garden after the first list and always return:
> 
> ```
> return []reconcile.Request{{NamespacedName: types.NamespacedName{Name: r.GardenName}}}
> ```

This would return the wrong name in case someone deletes a Garden and creates it again using a different name.